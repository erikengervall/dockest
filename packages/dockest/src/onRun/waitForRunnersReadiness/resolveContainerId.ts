import { Runner } from '../../runners/@types'
import DockestError from '../../errors/DockestError'
import execaWrapper from '../../utils/execaWrapper'
import sleep from '../../utils/sleep'

const getContainerId = async (runner: Runner): Promise<string> => {
  const {
    runnerConfig: { service },
  } = runner
  const command = ` \
                docker ps \
                  --quiet \
                  --filter \
                  "name=${service}" \
                --latest \
              `

  const containerId = await execaWrapper(command, runner)

  return containerId
}

const testables = { getContainerId }
export { testables }

export default async (runner: Runner): Promise<void> => {
  const {
    logger,
    runnerConfig: { service },
  } = runner
  const resolveContainerIdTimeout = 10
  let containerId = ''

  const recurse = async (resolveContainerIdTimeout: number): Promise<void> => {
    logger.info('Resolving containerId')

    if (resolveContainerIdTimeout <= 0) {
      throw new DockestError(
        `${service}: Timed out (${resolveContainerIdTimeout}s) while trying to resolve containerId`,
      )
    }

    try {
      containerId = await getContainerId(runner)

      if (typeof containerId !== 'string' || (typeof containerId === 'string' && containerId.length === 0)) {
        throw new Error(`Invalid containerId: ${containerId}`)
      }

      logger.info(`Resolved containerId: ${containerId}`)
    } catch (error) {
      resolveContainerIdTimeout--

      await sleep(1000)
      await recurse(resolveContainerIdTimeout)
    }

    runner.containerId = containerId
  }

  await recurse(resolveContainerIdTimeout)
}
