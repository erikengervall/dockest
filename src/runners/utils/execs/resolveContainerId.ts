import { DockestError } from '../../../errors'
import { runnerLogger } from '../../../loggers'
import { RunnerConfigs } from '../../index'
import { getContainerId, sleep } from '../index'

const resolveContainerId = async (runnerConfig: RunnerConfigs): Promise<string> => {
  const { service } = runnerConfig
  const timeout = 30

  let containerId = ''
  const recurse = async (timeout: number): Promise<string> => {
    runnerLogger.resolveContainerId(service)

    if (timeout <= 0) {
      throw new DockestError(`${service} getContainerId timed out`)
    }

    try {
      containerId = await getContainerId(service)

      if (
        typeof containerId !== 'string' ||
        (typeof containerId === 'string' && containerId.length === 0)
      ) {
        throw new Error('no bueno')
      }

      runnerLogger.resolveContainerIdSuccess(service, containerId)
    } catch (error) {
      timeout--

      await sleep(1000)
      await recurse(timeout)
    }

    return containerId
  }

  return recurse(timeout)
}

export default resolveContainerId
