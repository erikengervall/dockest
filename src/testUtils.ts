import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
