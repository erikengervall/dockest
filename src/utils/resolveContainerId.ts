import { DockestError } from '../errors'
import { Runner } from '../runners/@types'
import { execaWrapper, sleep } from './index'

const resolveContainerId = async (runner: Runner): Promise<void> => {
  const {
    runnerLogger,
    runnerConfig: { service },
  } = runner
  const resolveContainerIdTimeout = 10
  let containerId = ''

  const recurse = async (resolveContainerIdTimeout: number): Promise<void> => {
    runnerLogger.resolveContainerId()

    if (resolveContainerIdTimeout <= 0) {
      throw new DockestError(
        `${service}: Timed out (${resolveContainerIdTimeout}s) while trying to resolve containerId`
      )
    }

    try {
      containerId = await getContainerId(runner)

      if (
        typeof containerId !== 'string' ||
        (typeof containerId === 'string' && containerId.length === 0)
      ) {
        throw new Error(`Invalid containerId: ${containerId}`)
      }

      runnerLogger.resolveContainerIdSuccess(containerId)
    } catch (error) {
      resolveContainerIdTimeout--

      await sleep(1000)
      await recurse(resolveContainerIdTimeout)
    }

    runner.containerId = containerId
  }

  await recurse(resolveContainerIdTimeout)
}

const getContainerId = async (runner: Runner): Promise<string> => {
  const {
    runnerConfig: { service },
  } = runner
  const command = ` \
                docker ps \
                  --quiet \
                  --filter \
                  "name=${service}" \
                --latest \
              `

  const containerId = await execaWrapper(command, runner)

  return containerId
}

const testables = { getContainerId }
export { testables }
export default resolveContainerId
