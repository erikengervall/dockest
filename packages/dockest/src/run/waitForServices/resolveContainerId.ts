import { DockestError } from '../../Errors'
import { execaWrapper } from '../../utils/execaWrapper'
import { GENERATED_COMPOSE_FILE_PATH } from '../../constants'
import { Runner } from '../../@types'
import { sleep } from '../../utils/sleep'

const logPrefix = '[Resolve Container Id]'

const getContainerId = async (runner: Runner) => {
  const {
    dockestService: { serviceName },
  } = runner

  const command = `docker-compose \
                    -f ${GENERATED_COMPOSE_FILE_PATH} \
                    ps \
                    -q \
                    "${serviceName}"`

  const { stdout } = await execaWrapper(command, { runner, logPrefix })

  return stdout
}

export const resolveContainerId = async (runner: Runner) => {
  const { logger } = runner
  const resolveContainerIdTimeout = 10
  let containerId = ''

  const recurse = async (resolveContainerIdTimeout: number) => {
    if (resolveContainerIdTimeout <= 0) {
      throw new DockestError('Timed out', { runner })
    }

    logger.debug(`${logPrefix} Timeout in ${resolveContainerIdTimeout}s`)

    try {
      containerId = await getContainerId(runner)

      if (typeof containerId !== 'string' || (typeof containerId === 'string' && containerId.length === 0)) {
        throw new DockestError(`Invalid containerId (${containerId})`, { runner })
      }

      logger.info(`${logPrefix} Success (${containerId})`, { success: true })
    } catch (error) {
      resolveContainerIdTimeout--

      await sleep(1000)
      await recurse(resolveContainerIdTimeout)
    }

    runner.containerId = containerId

    return containerId
  }

  return await recurse(resolveContainerIdTimeout)
}
