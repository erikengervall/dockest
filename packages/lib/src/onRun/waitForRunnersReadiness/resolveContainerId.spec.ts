import { testables } from './resolveContainerId'
import testUtils, { mockedExecaStdout } from '../../testUtils'

const { getContainerId } = testables
const {
  initializedRunners: { redisRunner },
  execa,
} = testUtils({})

jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))

describe('getContainerId', () => {
  it('should work', async () => {
    const containerId = await getContainerId(redisRunner)

    expect(redisRunner.logger.debug).toMatchSnapshot()
    expect(execa).toMatchSnapshot()
    expect(containerId).toMatchSnapshot()
  })
})
