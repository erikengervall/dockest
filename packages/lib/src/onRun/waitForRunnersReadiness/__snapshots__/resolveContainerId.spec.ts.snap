// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getContainerId should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Executing shell script: docker ps --quiet --filter \\"name=redis\\" --latest",
    ],
    Array [
      "Successfully executed shell script docker ps --quiet --filter \\"name=redis\\" --latest",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`getContainerId should work 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "docker ps --quiet --filter \\"name=redis\\" --latest",
      Object {
        "shell": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "stdout": "getContainerId ðŸŒ®",
      },
    },
  ],
}
`;

exports[`getContainerId should work 3`] = `"getContainerId ðŸŒ®"`;
